plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/conversions/RemoveUnnecessaryParenthesesConversion.kt

```diff
@@ -0,0 +1,213 @@
+// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
+
+package org.jetbrains.kotlin.nj2k.conversions
+
+import org.jetbrains.kotlin.analysis.api.KtAnalysisSession
+import org.jetbrains.kotlin.nj2k.NewJ2kConverterContext
+import org.jetbrains.kotlin.nj2k.RecursiveConversion
+import org.jetbrains.kotlin.nj2k.identifier
+import org.jetbrains.kotlin.nj2k.symbols.JKUniverseClassSymbol
+import org.jetbrains.kotlin.nj2k.symbols.isStaticMember
+import org.jetbrains.kotlin.nj2k.tree.*
+import org.jetbrains.kotlin.nj2k.types.fqName
+import org.jetbrains.kotlin.parsing.KotlinExpressionParsing
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+import org.jetbrains.kotlin.utils.filterIsInstanceAnd
+
+/**
+ * Analogous to `RemoveUnnecessaryParenthesesIntention`.
+ * Removes parentheses (conservatively) added by other steps.
+ */
+class RemoveUnnecessaryParenthesesConversion(context: NewJ2kConverterContext) : RecursiveConversion(context) {
```

Comment:
Have you considered the opposite approach, i.e., do a call site check before generating a set of parentheses (the check can possibly be performed inside `org.jetbrains.kotlin.nj2k.ExpressionsKt#parenthesize`). In this case, any following conversions won't be able to circumvent this conversion and accidentally introduce redundant parentheses. But I understand that it can be harder to do an "in place" check while we are modifying the tree and maybe have incomplete information.

Reply:
I did, but like you said it's difficult/unwieldy for 2 reasons:
1. Like you mentioned, while we're building the JKTree, we don't have all the information we need about parent and descendent nodes
2. Sometimes parentheses become unnecessary in a later step (e.g. `((int) x).dec()` -> `(x as Int).dec()` -> `x.toInt().dec()`). We could require that every step take into consideration that it might need to remove now-unnecessary parentheses, but that seems laborious and error-prone, or at least not strictly better than this approach, IMO